### ISSUE TITLE
Python bindings fail to build

### ISSUE THREAD

**Issue by ixjlyons** [opensim-org/opensim-core#1257]
On Arch Linux, following the [using Makefiles](https://github.com/opensim-org/opensim-core#on-ubuntu-using-unix-makefiles) instructions, I found that the Python bindings failed to build:

```
make[2]: *** No rule to make target '../../Bindings/Python/Bindings/preliminaries.i', needed by 'Bindings/Python/actuators.py'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:6408: Bindings/Python/CMakeFiles/_actuators.dir/all] Error 2
make: *** [Makefile:139: all] Error 2
```

I fixed this by adding the following to `Bindings/Python/CMakeLists.txt`:

``` cmake
OpenSimAddPythonModule(MODULE preliminaries)
```

This pushed along the build process for a bit, but then I got a new error:

```
No module name specified using %module or -module.
make[2]: *** [Bindings/Python/CMakeFiles/_preliminaries.dir/build.make:64: Bindings/Python/preliminaries.py] Error 1
make[1]: *** [CMakeFiles/Makefile2:6705: Bindings/Python/CMakeFiles/_preliminaries.dir/all] Error 2
make: *** [Makefile:139: all] Error 2
```

Following the advice [here](https://stackoverflow.com/questions/12139093/swig-no-module-name-specified-using-module-or-module), I added this to `python_preliminaries.i`:

``` swig
%module(directors="1") preliminaries
```

Finally, the build completed successfully.

I'm completely new to OpenSim, so I don't really have a good grasp on how everything is set up yet. I am curious about how the Travis build is passing though.

**Comment by chrisdembia** [comment]
Hey Kenneth; I'm happy you were able to eventually successfully build the wrapping. Thank you for reporting this issue. There should not be a separate `preliminaries` module. The target `Bindings/Python/actuators.py` depends on the file `preliminaries.i`. This file is located in the source code at `Bindings/preliminaries.i`. However, Make seems to think that this file is located at `../../Bindings/Python/Bindings/preliminaries.i`; this path is incorrect and it's confusing to me why Make thinks that's where the file is.

There are two potential causes that I can think of:

1. We automatically generate the dependencies for the SWIG-related targets, by invoking SWIG during CMake's configure stage. Perhaps there is a bug in this CMake script. See [here](https://github.com/opensim-org/opensim-core/blob/master/Bindings/Python/CMakeLists.txt#L93).
2. You invoked `make` from somewhere other than the root of the build directory, which led to some path confusion.

To help us figure out what caused the issue, we need some more information.

1. What version of SWIG did you use?
2. What was the full path of your source directory, and what was the full path of your build directory?

**Comment by chrisdembia** [comment]
> I'm completely new to OpenSim, so I don't really have a good grasp on how everything is set up yet.

You seem to have a good grasp! You were able to play around with the CMake files.

> I am curious about how the Travis build is passing though.

Maybe because Travis uses Ubuntu? Looking at the Travis script, do you see any ways in which your process may have deviated?

**Comment by ixjlyons** [comment]
One difference is that I wasn't creating separate directories for *everything*. I tend to make a `build` directory within the root dir of the repository (hence the relative paths in the error I'm guessing). I tried again by setting up a little "opensim environment" in `~/tmp/opensim` and replaced all occurrences of `~` in the installation instructions with `~/tmp/opensim`.

Here's what I did step by step:

```bash
cd ~/tmp/opensim
git clone git@github.com:opensim-org/opensim-core.git
cd ~/tmp/opensim/opensim-core
cmake-gui
### configure the dependencies build with
        ~/tmp/opensim/opensim_dependencies-build
        ~/tmp/opensim/opensim_dependencies_install
cd ~/tmp/opensim_dependencies-build
make -j3
cd ~/tmp/opensim/opensim-core
cmake-gui
### configure the core build with
        ~/tmp/opensim/opensim_core-build
        ~/tmp/opensim/opensim_core_install
cd ~/tmp/opensim/opensim_core-build
make doxygen
make -j3
```

Still failing with the message related to `preliminaries.i`, this time with absolute paths:

```
make[2]: *** No rule to make target '/home/kenny/tmp/opensim/opensim-core/Bindings/Python/Bindings/preliminaries.i', needed by 'Bindings/Python/actuators.py'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:1776: Bindings/Python/CMakeFiles/_actuators.dir/all] Error 2
```

Arch is rolling release, so I have very recent versions of everything. In some cases this can be less convenient than it sounds:

```
GNU Make 4.2.1
gcc version 6.2.1 20160830 (GCC) 
SWIG Version 3.0.10
```

Here's the core build [CMakeCache.txt](https://github.com/opensim-org/opensim-core/files/477961/CMakeCache.txt)

**Comment by maxime-tournier** [comment]
Hello,

Just to mention that I had the same issue, also building from the root source directory. 

Building from a separate directory fixed the issue (you still need to delete leftover `CMakeCache.txt` in the root source dir though).

Hope this helps,

**Comment by chrisdembia** [comment]
Thank you @maxime-tournier for reporting. I would discourage using the source directory as the build directory. Still, if you submit a PR that fixes your issue, I would review it.

**Comment by syllogismos** [comment]
Hey, I followed these instructions to install opensim https://github.com/opensim-org/opensim-core#on-mac-osx-using-xcode but my python bindings are not working. When I run `ctest -R python -V` this is the out put I'm getting..

```
$ ctest -R python -V
UpdateCTestConfiguration  from :/Users/anil/Code/ai/crowdai/opensim/opensim_build/DartConfiguration.tcl
Parse Config file:/Users/anil/Code/ai/crowdai/opensim/opensim_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/Users/anil/Code/ai/crowdai/opensim/opensim_build/DartConfiguration.tcl
Parse Config file:/Users/anil/Code/ai/crowdai/opensim/opensim_build/DartConfiguration.tcl
Test project /Users/anil/Code/ai/crowdai/opensim/opensim_build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 84
    Start 84: python_tests

84: Test command: /usr/local/bin/python2.7 "-m" "unittest" "discover" "--start-directory" "/Users/anil/Code/ai/crowdai/opensim/opensim-core/Bindings/Python/tests" "--verbose"
84: Test timeout computed to be: 1500
84: Fatal Python error: PyThreadState_Get: no current thread
1/2 Test #84: python_tests .....................***Exception: Other  0.14 sec
test 85
    Start 85: python_examples

85: Test command: /usr/local/bin/python2.7 "-m" "unittest" "discover" "--start-directory" "/Users/anil/Code/ai/crowdai/opensim/opensim-core/Bindings/Python/examples" "--pattern" "*.py" "--verbose"
85: Test timeout computed to be: 1500
85: Fatal Python error: PyThreadState_Get: no current thread
2/2 Test #85: python_examples ..................***Exception: Other  0.15 sec

0% tests passed, 2 tests failed out of 2

Total Test time (real) =   0.30 sec

The following tests FAILED:
	 84 - python_tests (OTHER_FAULT)
	 85 - python_examples (OTHER_FAULT)
Errors while running CTest
```
Along with an osx warning, saying `python quit unexpectedly` warning like below
<img width="994" alt="screen shot 2017-03-05 at 22 41 33" src="https://cloud.githubusercontent.com/assets/929833/23589547/05a20aaa-01f5-11e7-8c9b-ac4b732ab69a.png">

**Comment by chrisdembia** [comment]
It's great that you are trying to use the python bindings. Thank you for the very thorough report.

It seems that you have homebrew python. The issue is almost certainly the issue described in the README: 
![image](https://cloud.githubusercontent.com/assets/846001/23590825/cbe6d38a-019b-11e7-97a9-015df85f225e.png).

Please make sure that the PYTHON_* cmake variables are all set consistently (to use the same exact python installation; either macOS's python or homebrew python). This is a bug with CMake I think.

**Comment by chrisdembia** [comment]
Closing for now. This might be fixed by #1610 .
